cdk/src/open-data-platform/data-plane/data-import/write-water-systems-data-handler.handler.ts

```diff
@@ -60,47 +73,58 @@ function parseS3IntoLeadServiceLinesTableRow(
       new Batch({ batchSize: batchSize }),
     ]);
 
+    const endIndex = startIndex + numberOfRowsToWrite;
+    let results: WaterSystemsTableRow[] = [];
     pipeline
-      .on('data', async (batch: any[]) => {
-        if (numberRows < numberOfRowsToWrite) {
-          const results: WaterSystemsTableRow[] = [];
-
-          batch.forEach((data) => {
-            const value = data.value;
+      .on('data', async (rows: any[]) => {
+        if (numberRowsParsed >= startIndex && numberRowsParsed < endIndex) {
+          for (let row of rows) {
+            const value = row.value;
             const properties = value.properties;
             const lead_connections =
-              properties.lead_connections != 'NaN' ? properties.lead_connections : 0;
+              properties.lead_connections != 'NaN' && properties.lead_connections != null
+                ? properties.lead_connections
+                : 0.0;
 
-            const row = new WaterSystemsTableRowBuilder()
+            const tableRowToInsert = new WaterSystemsTableRowBuilder()
               .pwsId(properties.pwsid)
               // Sometimes this number is negative because it is based on a
               // regression.
               .leadConnectionsCount(Math.max(parseFloat(lead_connections), 0))
               // Keep JSON formatting. Post-GIS helpers depend on this.
               .geom(JSON.stringify(value.geometry))
               .build();
-            results.push(row);
-          });
-          numberRows += batch.length;
-
-          // Pause reads while inserting into db.
-          pipeline.pause();
-          await insertRows(db, results);
-          pipeline.resume();
-        } else {
+            results.push(tableRowToInsert);
+          }
+          // Every batch size, write into the db.
+          if (results.length == batchSize) {
+            promises.push(insertRows(db, results));
+            results = [];
+          }
+        } else if (numberRowsParsed >= endIndex) {
+          // If there are any results left, write those.
+          if (results.length > 0) {
+            promises.push(insertRows(db, results));
+          }
+
           // Stop reading stream if numberOfRowsToWrite has been met.
           pipeline.destroy();
         }
+        numberRowsParsed += rows.length;
+        console.log(`Parsed ${numberRowsParsed}`);
       })
-      .on('error', (error: Error) => {
+      .on('error', async (error: Error) => {
+        await Promise.all(promises);
         reject(error);
       })
       // Gets called by pipeline.destroy()
       .on('close', async (_: Error) => {
-        resolve(numberRows);
+        await Promise.all(promises);
+        resolve(numberRowsParsed);
       })
       .on('end', async () => {
-        resolve(numberRows);
+        await Promise.all(promises);
```

Comment:
does this still timeout?

Reply:
No this works now to load the whole file in 2 runs